File: src/tools/llmObservability/getLLMCosts.ts
import type { z } from "zod";
import type { Context, Tool } from "../types";
import { LLMObservabilityGetCostsSchema } from "../../schema/tool-inputs";

const schema = LLMObservabilityGetCostsSchema;

type Params = z.infer<typeof schema>;

export const getLLMCostsHandler = async (context: Context, params: Params) => {
	const { projectId, days } = params;

	const trendsQuery = {
		kind: "TrendsQuery",
		dateRange: {
			date_from: `-${days || 6}d`,
			date_to: null,
		},
		filterTestAccounts: true,
		series: [
			{
				event: "$ai_generation",
				name: "$ai_generation",
				math: "sum",
				math_property: "$ai_total_cost_usd",
				kind: "EventsNode",
			},
		],
		breakdownFilter: {
			breakdown_type: "event",
			breakdown: "$ai_model",
		},
	};

	const costsResult = await context.api.query({ projectId }).execute({ queryBody: trendsQuery });
	if (!costsResult.success) {
		throw new Error(`Failed to get LLM costs: ${costsResult.error.message}`);
	}
	return {
		content: [{ type: "text", text: JSON.stringify(costsResult.data.results) }],
	};
};

const tool = (): Tool<typeof schema> => ({
	name: "get-llm-total-costs-for-project",
	description: `
        - Fetches the total LLM daily costs for each model for a project over a given number of days.
        - If no number of days is provided, it defaults to 7.
        - The results are sorted by model name.
        - The total cost is rounded to 4 decimal places.
        - The query is executed against the project's data warehouse.
        - Show the results as a Markdown formatted table with the following information for each model:
            - Model name
            - Total cost in USD
            - Each day's date
            - Each day's cost in USD
        - Write in bold the model name with the highest total cost.
        - Properly render the markdown table in the response.
    `,
	schema,
	handler: getLLMCostsHandler,
});

export default tool;

File: python/tools/llm_observability/get_llm_costs.py
import json

from api.client import is_error, is_success
from schema.tool_inputs import LLMObservabilityGetCostsSchema
from tools.tool_definitions import get_tool_definition
from tools.types import Context, Tool, ToolResult


async def get_llm_costs_handler(context: Context, params: LLMObservabilityGetCostsSchema) -> ToolResult:
    days = int(params.days) if params.days else 6

    trends_query = {
        "kind": "TrendsQuery",
        "dateRange": {"date_from": f"-{days}d", "date_to": None},
        "filterTestAccounts": True,
        "series": [
            {
                "event": "$ai_generation",
                "name": "$ai_generation",
                "math": "sum",
                "math_property": "$ai_total_cost_usd",
                "kind": "EventsNode",
            }
        ],
        "breakdownFilter": {"breakdown_type": "event", "breakdown": "$ai_model"},
    }

    costs_result = await context.api.query(str(params.projectId)).execute({"query": trends_query})
    if is_error(costs_result):
        raise Exception(f"Failed to get LLM costs: {costs_result.error}")

    assert is_success(costs_result)

    return ToolResult(content=json.dumps(costs_result.data.results))


def get_llm_costs_tool() -> Tool[LLMObservabilityGetCostsSchema]:
    definition = get_tool_definition("get-llm-total-costs-for-project")

    return Tool(
        name="get-llm-total-costs-for-project",
        description=definition["description"],
        schema=LLMObservabilityGetCostsSchema,
        handler=get_llm_costs_handler,
    )
